[Login Feature] Added Login/Create Account Views in Home Page (FRONT END ONLY)

fnormand01 opened the pull request:
# Description

I implemented some new authentication components, meaning one for the login and one for the account creation.

**Login:**
There is no validation at all in this component. As soon as you click _Login_, you'll be redirected to the home page with a welcome message. Basic authentication logic has been used to produce this _fake_ login behavior. This will be removed once we'll introduce the back end implementation of the login feature.

Note that as soon as you refresh or change the page, the login process will restart, meaning that you don't stay logged in at the moment.

**Created Account:**
In this component, there is a lot of validation. All fields are validated using RegEx. Also, the _Create account_ button stay disabled as long as all the validation are not satisfied. We'll have to implement the _Submit_ data in the back end in the future for this component.

## Type of change

- New feature (non-breaking change which adds functionality)

## Screenshot

**Login Page:**

![image](https://user-images.githubusercontent.com/23021242/55151950-b98a5000-5125-11e9-93b7-1a3c6173720d.png)

**As soon as you click _Login_:**

![image](https://user-images.githubusercontent.com/23021242/55152095-09691700-5126-11e9-958f-36e1a15ec982.png)

**Create Account Page:**

![image](https://user-images.githubusercontent.com/23021242/55152154-24d42200-5126-11e9-88b6-c8cc97656135.png)

**Validation Examples:**

_First Name contains a special character_
![image](https://user-images.githubusercontent.com/23021242/55152240-53ea9380-5126-11e9-88ee-f522066ae50c.png)

_Missing '@' in Email Address_
![image](https://user-images.githubusercontent.com/23021242/55152322-83999b80-5126-11e9-88ec-2177742be0a3.png)

_Password that doesn't satisfy all the requirements_
![image](https://user-images.githubusercontent.com/23021242/55152378-a4fa8780-5126-11e9-883d-dbca63d4d697.png)

_Password Confirmation that isn't the same as the Password_
![image](https://user-images.githubusercontent.com/23021242/55152417-bd6aa200-5126-11e9-8526-14162089c505.png)

**When every validations are satisfied (the button becomes enabled):**

![image](https://user-images.githubusercontent.com/23021242/55152518-f0ad3100-5126-11e9-8b88-c91ba70617e2.png)

# Testing

Manual steps to reproduce this functionality:

1.  Go to _Home_ page.
2.  Play with the _Login_ and _Create Account_ forms and make sure everything works/looks good.

**Note that only front end implementation has been done, so there is no submission, logging control, logout functionality, etc.**

# Checklist

Applicable for all code changes.

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] ~~I have made corresponding changes to the documentation~~
- [x] My changes generate no new compiler warnings
- [ ] ~~My changes generate no new accessibility errors and/or warnings~~
- [ ] ~~I have added tests that prove my fix is effective or that my feature works~~
- [x] I have researched WCAG2.1 accessibility standards and met them in this PR (can be navigated using a keyboard)
- [x] My changes look good on IE 10+, Firefox, and Chrome

caleybrock commented on frontend/src/components/authentication/AuthenticationTabs.jsx:
```diff
@@ -0,0 +1,73 @@
+              tabSpecs={TABS}
+              currentTab={0}
+              menuName={LOCALIZE.ariaLabel.authenticationMenu}
+              style={{
```
This object should be defined in custom styles, then referred to here. `style={customStyles.tabNavigation]`

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/0b7d37807d5359a65b89fa50e4435a6b2649ccbd

caleybrock commented on frontend/src/components/authentication/CreateAccountForm.jsx:
```diff
@@ -0,0 +1,310 @@
+
+let SUBMIT_BTN_DISABLED;
+
+class LoginForm extends Component {
```
The class names you are using do not match the name of the file.

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/c69b9a2f0170c5941645d863a72f8391adf869cb

caleybrock commented on frontend/src/components/authentication/CreateAccountForm.jsx:
```diff
@@ -0,0 +1,310 @@
+  }
+};
+
+let SUBMIT_BTN_DISABLED;
```
This should be defined and controlled in the component state. It's a best practice to not have global variables and if the component is used multiple times it resets itself properly.

fnormand01 replied:
I put a comment about that on line 136:

![image](https://user-images.githubusercontent.com/23021242/55161592-47713580-513c-11e9-9a8e-6403e8203004.png)

If you know how to fix that, I'd like to know how.

caleybrock commented on frontend/src/components/authentication/CreateAccountForm.jsx:
```diff
@@ -0,0 +1,310 @@
+  }
+}
+
+export default LoginForm;
```
Name doesn't match here either.

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/c69b9a2f0170c5941645d863a72f8391adf869cb

caleybrock commented on frontend/src/text_resources.js:
```diff
@@ -14,7 +14,52 @@ let LOCALIZE = new LocalizedStrings({
+            passwordPlaceholder: "Password",
+            passwordErrors: {
+              description: "Your password must satisfy the following:",
+              Bullet1: "At least one upper case",
```
The names of the strings should be more descriptive of the of what they are. Instead of `Bullet1`, maybe `uppercase`?

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/013f93506b747359bafe5c5049c1eec7587eb71d

caleybrock commented on frontend/src/helpers/regexValidator.js:
```diff
@@ -0,0 +1,39 @@
+function validateName(string) {
```
All of these functions are perfect candidates for unit tests - easy to check for expected input / output of expected cases.

fnormand01 replied:
I'll add tests for these new components in another PR, because this PR is already big enough.

caleybrock commented on frontend/src/components/eMIB/EmibTabs.jsx:
```diff
@@ -40,7 +40,19 @@ class EmibTabs extends Component {
+          currentTab={1}
+          menuName={LOCALIZE.ariaLabel.tabMenu}
+          style={{
+            height: "calc(100vh - 241px)",
```
This object should be defined in a styles constant.

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/8a6f6b3854fe17c2ebab4738ef62901e5ddda972

MichaelCherry commented on frontend/src/components/authentication/AuthenticationTabs.jsx:
```diff
@@ -0,0 +1,73 @@
+    const TABS = [
+      {
+        id: 0,
+        tabName: LOCALIZE.homePage.login.title,
```
(Comment Ref 1) Since the login won't live on the home page necessarily, I'm thinking this should be 'LOCALIZE.login' and then maybe have '.autheniticationTabs', '.createAccount' '.loginForm' underneath it

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/6ce81fd2552bf6633dd7149f12e43f64df4e3db2

MichaelCherry commented on frontend/src/components/authentication/LoginForm.jsx:
```diff
@@ -0,0 +1,95 @@
+        {!this.state.isAuthenticated && (
+          <div>
+            <div style={styles.loginContent}>
+              <h3>{LOCALIZE.homePage.login.content.title}</h3>
```
See (Comment Ref 1)

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/6ce81fd2552bf6633dd7149f12e43f64df4e3db2

MichaelCherry commented on frontend/src/components/authentication/CreateAccountForm.jsx:
```diff
@@ -0,0 +1,310 @@
+      <div>
+        <div>
+          <div style={styles.createAccountContent}>
+            <h3>{LOCALIZE.homePage.createAccount.content.title}</h3>
```
See (comment Ref 1)

fnormand01 replied:
Fixed in commit #https://github.com/code-for-canada/project-thundercat/pull/152/commits/6ce81fd2552bf6633dd7149f12e43f64df4e3db2

MichaelCherry commented on frontend/src/components/authentication/AuthenticationTabs.jsx:
```diff
@@ -0,0 +1,73 @@
+              tabSpecs={TABS}
+              currentTab={0}
+              menuName={LOCALIZE.ariaLabel.authenticationMenu}
+              style={{
```
I'll resolve this one, since Caley already commented and this is a duplicate of that

fnormand01 commented:
We don't need this PR anymore, since I separated these changes in three PRs.

- Introducing Fields Validation (PR https://github.com/code-for-canada/project-thundercat/pull/154)
- Added new prop in _TabNavigation.jsx_ for style customization (PR https://github.com/code-for-canada/project-thundercat/pull/155)
- Added Login and Create Account Fforms + Included Validation (PR https://github.com/code-for-canada/project-thundercat/pull/156)
